[PL-31424] New topic for audit streaming. (#933)
harness/developer-hub: docs/platform/15_Audit-Trail/audit-streaming.md

- You can add multiple streaming destinations in Harness.

## Add a streaming destination

You can configure the audit stream from the **Audit Log Streaming** option in **Audit Trail**.

1. In your Harness account, select **Account Settings**.

2. Select **Audit Trail** and then select **Audit Log Streaming**.

3. Select **New Streaming Destination** to open the configuration settings.

4. Enter a name for the streaming destination.

5. Harness automatically populates the Id for the destination. You can choose to retain the Id or modify it.

6. Enter a description and tag(s) for the streaming destination.

7. Select **Continue**.
   
![](./static/audit-streaming.png)

## Configure the streaming connector

1. Select **Amazon S3**.

2. In **Select Connector**, select an existing AWS Cloud Provider connector or create a new one.

   Go to [Add an AWS connector](../7_Connectors/add-aws-connector.md) for steps to create a new AWS Cloud Provider connector.

3. Select **Apply Selected**.

4. In **Amazon S3 Bucket**, enter the bucket name.
   
   This is where Harness will write all the streaming records.

   ![](./static/streaming-connector.png)

5. Select Save and Continue.

> 6. After the connection test is successful, select **Finish**.

Reviewer: The test connection currently doesn't account the bucket used.Also it is doing test connection for the connector used and not doing the actual test connection for streaming destination i.e. the bucket in AWS.

> 6. After the connection test is successful, select **Finish**.

Reviewer: The test connection currently doesn't account the bucket used.Also it is doing test connection for the connector used and not doing the actual test connection for streaming destination i.e. the bucket in AWS.

## Activate or deactivate streaming

1. After you configure the streaming destination, it appears in the list of destinations in **Audit Log Streaming**.
   
   By default, this destination is inactive.

2. To activate streaming to this destination, toggle the status to **Active** besides the streaming destination.

   After you activate the destination, Harness starts writing audit logs in the destination.

   Post that every 30 mins audit logs are streamed.

> 3. You can pause audit streaming, preventing any new audit events from being streamed to the configured endpoint by setting the status to **Inactive**. 

Reviewer: Currently it will start streaming from where it was paused. In future we are going to limit it to a threshold value. As an e.g. it can be paused (inactive) for a max of duration of X hrs. If it is marked active within X hrs from marking inactive then the streaming will start from where it was paused. else it will start for the time its marked active.

## Update audit stream

You can change the audit stream configuration by clicking three dots beside the stream destination. This will open a pop-up menu with the following options:

- Edit: You can select a different streaming destination or make changes to the configured destination.

> - Delete: You can delete the audit stream destination, which will clear any data relating to the audit stream and prevent any new audit events from being streamed.

Reviewer: Th history of batched is currently not deleted. So if a used creates a streaming destination again with same identifier then it will pick the state which is already present in db. We need to discuss and close on the behaviour of delete.

---
title: Configure Audit Streaming
description: Stream your audit logs to an external destination.
sidebar_position: 3
---

:::note
Currently, this feature is in Beta and behind the feature flag `PL_AUDIT_LOG_STREAMING_ENABLED`. Contact Harness Support to enable the feature.
:::

You can configure a streaming destination in Harness to send audit log data to another location for processing. Integrating audit data with other Security Incident and Event Management (SIEM) tools lets you do the following:

- Trigger alerts for specific events.

- Create views of audit data.

- Perform anomaly detection. 

> - Store more than 2 years worth of auditing data, which is the maximum amount of data that Harness keeps for your accounts.

Reviewer 2: ```suggestion
- Store more than 2 years audit data, which is the maximum amount of data that Harness keeps for your accounts.
```

- Keep your resources and data protected, and ensure compliance with regulations.

Harness streams the audit events to your chosen SIEM tool as structured JSON.

## Important

- Streaming destinations receive all audit event data, which could include sensitive information. Make sure you trust the streaming destination.
- Harness supports streaming to Amazon S3.
- You can add multiple streaming destinations in Harness.

## Add a streaming destination

> You can configure the audit stream from the **Audit Log Streaming** option in **Audit Trail**.

Reviewer 2: As this is part of the steps, can we just say "To add a streaming destination in Harness:"

1. In your Harness account, select **Account Settings**.

2. Select **Audit Trail** and then select **Audit Log Streaming**.

3. Select **New Streaming Destination** to open the configuration settings.

4. Enter a name for the streaming destination.

> 5. Harness automatically populates the Id for the destination. You can choose to retain the Id or modify it.

Reviewer 2: Harness populates the destination's Id automatically. You can retain or change the Id.

6. Enter a description and tag(s) for the streaming destination.

7. Select **Continue**.
   
![](./static/audit-streaming.png)

## Configure the streaming connector

1. Select **Amazon S3**.

2. In **Select Connector**, select an existing AWS Cloud Provider connector or create a new one.

   Go to [Add an AWS connector](../7_Connectors/add-aws-connector.md) for steps to create a new AWS Cloud Provider connector.

3. Select **Apply Selected**.

4. In **Amazon S3 Bucket**, enter the bucket name.
   

>    This is where Harness will write all the streaming records.

Reviewer 2: Harness writes all the streaming records to this destination.

   ![](./static/streaming-connector.png)

> 5. Select Save and Continue.

Reviewer 2: ```suggestion
5. Select **Save** and **Continue**.
```

6. After the connection test is successful, select **Finish**.
   
   :::note

>    AWS buckets are not taken into account during the test connection. Also, it tests the connector used without testing the bucket.

Reviewer 2: Harness does not consider AWS buckets while testing the connection. Only the connector is tested.

   :::
   

## Activate or deactivate streaming

> 1. After you configure the streaming destination, it appears in the list of destinations in **Audit Log Streaming**.

Reviewer 2: "The streaming destination gets configured and appears in the list of destinations under **Audit Log Streaming**. By default setting of this destination is inactive."

Please move this to Step 6 above (this is the result of that step).

   
   By default, this destination is inactive.

> 2. To activate streaming to this destination, toggle the status to **Active** besides the streaming destination.

Reviewer 2: Please consider if you can make this section a single paragraph instead of a list.

>    After you activate the destination, Harness starts writing audit logs in the destination.

Reviewer 2: Harness starts writing audit logs after the destination is activated.

>    Post that every 30 mins audit logs are streamed.

Reviewer 2: Harness streams audit logs to destination at 30-minute intervals.

> 3. You can pause audit streaming, preventing any new audit events from being streamed to the configured endpoint by setting the status to **Inactive**. 

Reviewer 2: By pausing audit streaming, you can prevent new audit events from being transmitted to the configured endpoint. To do this, set the streaming destination status to **Inactive**.

Author: I will retain what I have written here. It is more about the available actions.

>    Harness streams audit logs from where it was paused later when you reactivate such a streaming destination.

Reviewer 2: When you reactivate the streaming destination, Harness starts streaming the audit logs from the point where it was paused.

## Update audit stream

> You can change the audit stream configuration by clicking three dots beside the stream destination. This will open a pop-up menu with the following options:

Reviewer 2: 1. stream destination or streaming destination? Please review.

2. A pop-up menu appears with the followoing options:

> - Edit: You can select a different streaming destination or make changes to the configured destination.

Reviewer 2: ```suggestion
- **Edit**: Select a different streaming destination or make changes to the existing destination.
```

> - Delete: You can delete the audit stream destination.

Reviewer 2: ```suggestion
- Delete: Delete the audit stream destination.
```
